File: swift-algos/calcEquation.swift


// build adjacency dictionary 


// then do bfs on the query until you arrive at the destination
// use bfs to minimize the traversal 



func calcEquation(_ equations: [[String]], _ values: [Double], _ queries: [[String]]) -> [Double] {


    // build adjacency dictionary 
    var graph = [String: [(String, Double)]]()

    for (i, coords) in equations.enumerated() {
        let u: String = coords[0]
        let v: String = coords[1]
        let val = values[i]
        if let _ = graph[u] {
            graph[u]!.append((v, val))
        } else {
            graph[u] = [(v, val)]
        }

        if let _ = graph[v] {
            graph[v]!.append((u, 1/val))
        } else {
            graph[v] = [(u, 1/val)]
        }
    }
    print(graph)


    // define bfs

    func bfs(_ u: String, _ v: String) -> Double {
        var queue: [String] = []
        var start: String = u





        return 1.0
    }


    // now for each query, bfs

    return [1, 2]
}


var equations = [["a","b"],["b","c"]]
var values = [2.0,3.0]
var queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]

var x = calcEquation(equations, values, queries)


var queue = Deque<String>  

File: python-fundamentals/graphs/calcEquation.py
from collections import defaultdict, deque

def calcEquation(equations, values, queries): 
    graph = defaultdict(list)
    for i, (u,v) in enumerate(equations): 
        val = values[i]
        graph[u].append((v, val))
        graph[v].append((u, 1/val))

    def bfs(start, end):
        if start not in graph: 
            return -1
        queue = deque([(start, 1, set([start]))])
        res = 1
        while queue: 
            u, u_val, u_visited = queue.popleft()
            if u == end: 
                return u_val
            u_visited = set(list(u_visited) + [u])
            for v, v_val in graph[u]:
                if v not in u_visited: 
                    
                    queue.append((v, u_val * v_val, u_visited))
        return -1
    
    res = []
    for u, v in queries: 
        res.append(bfs(u, v))

    return res

equations = [["a","b"],["b","c"]]
values = [2.0,3.0]
queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]

print(calcEquation(equations, values, queries))

# queue = deque(["1", "2"])
# print(queue.popleft())
# print(queue.popleft())
# # print(queue.popleft())

# v_visited = set([1,2]) 
# v_visited = list(v_visited) + [3]
# # print(v_visited)
# # print()